gliner/trainer.py

```diff
@@ -0,0 +1,256 @@
+from transformers import get_cosine_schedule_with_warmup
+
+def divisible_by(num, den):
+    return (num % den) == 0
+
+def cycle(dl):
+    while True:
+        for data in dl:
+            yield data
+
+def exists(v):
+    return v is not None
+
+def default(v, d):
+    return v if exists(v) else d
+
+DEFAULT_DDP_KWARGS = DistributedDataParallelKwargs(
+    find_unused_parameters = True
+) 
+class GlinerTrainer(Module):
+    @beartype
+    def __init__(
+        self,
+        model: GLiNER,
+        train_data: list, 
+        batch_size: int,
+        
+        grad_accum_every: int | None = None, 
+        max_types: int = 25, # maximum number of entity types during training
+        shuffle_types: bool = True, # if shuffle or not entity types
+        random_drop: bool = True, # randomly drop entity types
+        max_neg_type_ratio: int = 1, # ratio of positive/negative types, 1 mean 50%/50%, 2 mean 33%/66%, 3 mean 25%/75% ...
+        max_len: int = 384, # maximum sentence length
+        lr_encoder: float = 1e-5,
+        lr_others: float = 5e-5, 
+        
+        freeze_token_rep: bool = False, 
+        val_data: dict | None = None,
+        val_every_step : Optional[int] = None,
+        accelerator_kwargs: dict = dict(),
+        optimizer_kwargs: dict = dict(),
+        checkpoint_every_step: Optional[int] = None,
+        checkpoint_every_epoch: Optional[int] = None,
+        checkpoint_folder = './checkpoints',  
+        use_wandb_tracking = False
+    ):
+        super().__init__()
+        self.use_wandb_tracking = use_wandb_tracking
+
+        if use_wandb_tracking:
+            accelerator_kwargs['log_with'] = 'wandb'
+   
+        if 'kwargs_handlers' not in accelerator_kwargs:
+            accelerator_kwargs['kwargs_handlers'] = [DEFAULT_DDP_KWARGS]
+
+        self.accelerator = Accelerator(**accelerator_kwargs)
+
+        self.model = model
+  
+        self.model.set_sampling_params(
+            max_types=max_types, 
+            shuffle_types=shuffle_types, 
+            random_drop=random_drop, 
+            max_neg_type_ratio=max_neg_type_ratio, 
+            max_len=max_len
+        )
+        
+        param_groups = [ 
+            {"params": model.rnn.parameters(), "lr": lr_others}, 
+            {"params": model.span_rep_layer.parameters(), "lr": lr_others},
+            {"params": model.prompt_rep_layer.parameters(), "lr": lr_others},
+        ]
+
+        if not freeze_token_rep:
+            # If token_rep_layer should not be frozen, add it to the optimizer with its learning rate
+            param_groups.append({"params": model.token_rep_layer.parameters(), "lr": lr_encoder})
+        else:
+            # If token_rep_layer should be frozen, explicitly set requires_grad to False for its parameters
+            for param in self.token_rep_layer.parameters():
+                param.requires_grad = False
+                
+        self.optimizer = torch.optim.AdamW(param_groups, **optimizer_kwargs) 
```

Comment:
> you can use directly `model.get_optimizer(self, lr_encoder, lr_others, freeze_token_rep=False)` instead of line `80-94`

The reason why I didn't was due to the optimizer_kwargs which is needed for fp16 training. 
I probably can just change the model.get_optimizer function to include a optimizer_kwargs param.

Reply:
ok, I see

Reply:
> ok, I see

Changed it per suggestion: https://github.com/urchade/GLiNER/pull/85/commits/89803a433e094a59233d1cde5127b5ea9d3bb4aa